tests/virt_autotest/virt_autotest_base.pm

```diff
@@ -131,29 +190,11 @@ sub run_test {
         return;
     }
 
-    my $script_output = $self->execute_script_run($test_cmd, $timeout);
-
-    if ($add_junit_log_flag eq "yes") {
-        $self->add_junit_log($script_output);
-    }
-
-    if ($upload_virt_log_flag eq "yes") {
-        upload_virt_logs($log_dir, $compressed_log_name);
-        virt_utils::upload_supportconfig_log;
-        $self->upload_coredumps;
-    }
-
-    if ($upload_guest_assets_flag eq "yes") {
-        record_info('Check UPLOAD_GUEST_ASSETS flag', 'This test should upload guest assets!');
-        $self->upload_guest_assets;
-    }
-
-    if ($assert_pattern) {
-        unless ($script_output =~ /$assert_pattern/m) {
-            assert_script_run("grep -E \"$assert_pattern\" $log_dir -r || zcat /tmp/$compressed_log_name.tar.gz | grep -aE \"$assert_pattern\"");
-        }
-    }
-
+    $self->save_test_configuration($assert_pattern, $add_junit_log_flag, $upload_virt_log_flag, $log_dir, $compressed_log_name, $upload_guest_assets_flag);
+    $self->execute_script_run($test_cmd, $timeout);
+    $self->post_execute_script_run;
+    $self->post_run_test($self->{script_output});
```

Comment:
$self->post_execute_script_run which has assertions should be put as the last step. Please move  $self->post_run_test($self->{script_output}) above it. By it, post_fail_hook can save  $self->post_run_test($self->{script_output}) also.

Reply:
$self->{scirpt_output} may change during assertion, and it has to be used by post_run_test. So post_run_test must be called afterwards instead of before.

Reply:
If your new assertion does what the original assert check does, and the original assert check is just done after

Reply:
You could have wrong junit log. The original if condition statements are just split into two parts. One happens firstly, the other one must happen after. And your suggestion will lead to unexpected logic, post_run_test must be called either in run_test or post_fail_hook to maintain correct test logic (pass or fail).

Reply:
I guess it is because what new assertion(post_execute_script_run) does is different from before now. Better to double confirm the difference here.  I can accept difference, just if  it can handle the  output and job/subtest status as well as before. You can ignore my previous suggestion about it, since not much benefit, and not easy to verify it can work well after changing order.

Reply:
The current one has been well thought. I think I already mentioned I do preserve logic as before.